Optimizing ByChained#findElement

alb-i986 opened the pull request:
- optimize complexity: instead of traversing the entire "tree"
  of partial WebElements found, only travers the left-most path
- in case the bys list is empty, throw an exception with a more meaningful message

Not sure if you like the

```
assert elem != null;
```

I seem to have some issues in running the tests. They are failing, even on master. I guess it must be due to my env (?).
OSX 10.10.2, Java 1.8.0_40
Also tried java 6 but it didnt seem to help

```
Compiling: //java/client/test/org/openqa/selenium:SmallTests as build/java/client/test/org/openqa/selenium/SmallTests.jar
mkdir -p build/java/client/test/org/openqa/selenium/SmallTests.jar_temp
Compiling 1 source file to /Users/alb/workspace/selenium/build/java/client/test/org/openqa/selenium/SmallTests.jar_temp
Building jar: /Users/alb/workspace/selenium/build/java/client/test/org/openqa/selenium/SmallTests.jar
rm -rf build/java/client/test/org/openqa/selenium/SmallTests.jar_temp
** Execute //java/client/test/org/openqa/selenium:SmallTests
** Execute //java/client/test/org/openqa/selenium:SmallTests:run
Testing: //java/client/test/org/openqa/selenium:SmallTests
mkdir -p build/test_logs
Listening for transport dt_socket at address: 5005
Running org.openqa.selenium.SmallTests
Exception in thread "Thread-20" org.openqa.selenium.WebDriverException: Thread safety error; this instance of WebDriver was constructed on thread main (id 1) and is being accessed by thread Thread-20 (id 35)This is not permitted and *will* cause undefined behaviour
Build info: version: '2.45.0', revision: '17dc63a', time: '2015-03-26 03:09:18'
System info: host: 'albs-MacBook-Air.local', ip: '192.168.0.25', os.name: 'Mac OS X', os.arch: 'x86_64', os.version: '10.10.2', java.version: '1.8.0_40'
Driver info: driver.version: unknown
    at org.openqa.selenium.support.ThreadGuard$WebDriverInvocationHandler.invoke(ThreadGuard.java:86)
    at com.sun.proxy.$Proxy22.findElement(Unknown Source)
    at org.openqa.selenium.support.ThreadGuardTest$1.run(ThreadGuardTest.java:44)
    at java.lang.Thread.run(Thread.java:745)
Tests run: 345, Failures: 0, Errors: 0, Time elapsed: 1.831 sec
** Invoke test_support (first_time)
** Invoke //java/client/test/org/openqa/selenium/lift:test:run (first_time)
** Invoke //java/client/test/org/openqa/selenium/lift:test
** Execute //java/client/test/org/openqa/selenium/lift:test:run
Testing: //java/client/test/org/openqa/selenium/lift:test
mkdir -p build/test_logs
Listening for transport dt_socket at address: 5005
Running org.openqa.selenium.lift.SmallTests
Tests run: 23, Failures: 0, Errors: 0, Time elapsed: 0.394 sec
** Invoke //java/client/test/org/openqa/selenium/support:SmallTests:run (first_time)
** Invoke //java/client/test/org/openqa/selenium/support:SmallTests (first_time)
** Invoke build/java/client/test/org/openqa/selenium/support/SmallTests.jar (first_time)
** Invoke //java/client/test/org/openqa/selenium/support:tests
** Invoke java/client/test/org/openqa/selenium/support/SmallTests.java (not_needed)
** Execute build/java/client/test/org/openqa/selenium/support/SmallTests.jar
rm -rf build/java/client/test/org/openqa/selenium/support/SmallTests.jar_temp
Exception `Errno::ENOENT' at org/jruby/RubyFile.java:1729 - No such file or directory - build/java/client/test/org/openqa/selenium/support/SmallTests.jar_temp
Compiling: //java/client/test/org/openqa/selenium/support:SmallTests as build/java/client/test/org/openqa/selenium/support/SmallTests.jar
mkdir -p build/java/client/test/org/openqa/selenium/support/SmallTests.jar_temp
Compiling 1 source file to /Users/alb/workspace/selenium/build/java/client/test/org/openqa/selenium/support/SmallTests.jar_temp
Building jar: /Users/alb/workspace/selenium/build/java/client/test/org/openqa/selenium/support/SmallTests.jar
rm -rf build/java/client/test/org/openqa/selenium/support/SmallTests.jar_temp
** Execute //java/client/test/org/openqa/selenium/support:SmallTests
** Execute //java/client/test/org/openqa/selenium/support:SmallTests:run
Testing: //java/client/test/org/openqa/selenium/support:SmallTests
mkdir -p build/test_logs
Listening for transport dt_socket at address: 5005
Running org.openqa.selenium.support.SmallTests
Tests run: 181, Failures: 0, Errors: 0, Time elapsed: 1.158 sec
Exception in thread "Thread-1" org.openqa.selenium.WebDriverException: Thread safety error; this instance of WebDriver was constructed on thread main (id 1) and is being accessed by thread Thread-1 (id 12)This is not permitted and *will* cause undefined behaviour
Build info: version: '2.45.0', revision: '17dc63a', time: '2015-03-26 03:09:18'
System info: host: 'albs-MacBook-Air.local', ip: '192.168.0.25', os.name: 'Mac OS X', os.arch: 'x86_64', os.version: '10.10.2', java.version: '1.8.0_40'
Driver info: driver.version: unknown
    at org.openqa.selenium.support.ThreadGuard$WebDriverInvocationHandler.invoke(ThreadGuard.java:86)
    at com.sun.proxy.$Proxy20.findElement(Unknown Source)
    at org.openqa.selenium.support.ThreadGuardTest$1.run(ThreadGuardTest.java:44)
    at java.lang.Thread.run(Thread.java:745)
** Invoke //java/client/test/org/openqa/selenium/support:LargeTests:run (first_time)
** Invoke //java/client/test/org/openqa/selenium/support:LargeTests (first_time)
** Invoke build/java/client/test/org/openqa/selenium/support/LargeTests.jar (first_time)
** Invoke //java/client/test/org/openqa/selenium/support:tests
** Invoke java/client/test/org/openqa/selenium/support/LargeTests.java (not_needed)
** Execute build/java/client/test/org/openqa/selenium/support/LargeTests.jar
rm -rf build/java/client/test/org/openqa/selenium/support/LargeTests.jar_temp
Exception `Errno::ENOENT' at org/jruby/RubyFile.java:1729 - No such file or directory - build/java/client/test/org/openqa/selenium/support/LargeTests.jar_temp
Compiling: //java/client/test/org/openqa/selenium/support:LargeTests as build/java/client/test/org/openqa/selenium/support/LargeTests.jar
mkdir -p build/java/client/test/org/openqa/selenium/support/LargeTests.jar_temp
Compiling 1 source file to /Users/alb/workspace/selenium/build/java/client/test/org/openqa/selenium/support/LargeTests.jar_temp
Building jar: /Users/alb/workspace/selenium/build/java/client/test/org/openqa/selenium/support/LargeTests.jar
rm -rf build/java/client/test/org/openqa/selenium/support/LargeTests.jar_temp
** Execute //java/client/test/org/openqa/selenium/support:LargeTests
** Execute //java/client/test/org/openqa/selenium/support:LargeTests:run
Testing: //java/client/test/org/openqa/selenium/support:LargeTests
mkdir -p build/test_logs
Listening for transport dt_socket at address: 5005
Running org.openqa.selenium.support.LargeTests
Mar 26, 2015 3:12:14 AM org.openqa.selenium.testing.JUnit4TestBase$1 starting
INFO: >>> Starting selectByVisibleTextShouldNormalizeSpaces(org.openqa.selenium.support.ui.SelectLargeTest)
2015-03-26 03:12:14.709:INFO:osjs.Server:jetty-7.x.y-SNAPSHOT
2015-03-26 03:12:14.744:INFO:osjsh.ContextHandler:started o.s.j.s.ServletContextHandler{/common,file:/Users/alb/workspace/selenium/common/src/web/}
2015-03-26 03:12:14.748:INFO:/common:Aliases are enabled
2015-03-26 03:12:14.749:INFO:osjsh.ContextHandler:started o.s.j.s.ServletContextHandler{/javascript,file:/Users/alb/workspace/selenium/javascript/}
2015-03-26 03:12:14.751:INFO:/javascript:Aliases are enabled
2015-03-26 03:12:14.752:INFO:osjsh.ContextHandler:started o.s.j.s.ServletContextHandler{/third_party/closure/goog,file:/Users/alb/workspace/selenium/third_party/closure/goog/}
2015-03-26 03:12:14.752:INFO:/third_party/closure/goog:Aliases are enabled
2015-03-26 03:12:14.752:INFO:osjsh.ContextHandler:started o.s.j.s.ServletContextHandler{/third_party/js,file:/Users/alb/workspace/selenium/third_party/js/}
2015-03-26 03:12:14.752:INFO:/third_party/js:Aliases are enabled
2015-03-26 03:12:14.778:INFO:osjs.AbstractConnector:Started SelectChannelConnector@0.0.0.0:34396
2015-03-26 03:12:14.996:INFO:osjus.SslContextFactory:Enabled Protocols [SSLv2Hello, TLSv1, TLSv1.1, TLSv1.2] of [SSLv2Hello, SSLv3, TLSv1, TLSv1.1, TLSv1.2]
2015-03-26 03:12:15.007:INFO:osjs.AbstractConnector:Started SslSocketConnector@0.0.0.0:20733
Tests run: 5, Failures: 0, Errors: 5, Time elapsed: 0.757 sec
Mar 26, 2015 3:12:15 AM org.openqa.selenium.testing.JUnit4TestBase$1 finished
INFO: <<< Finished selectByVisibleTextShouldNormalizeSpaces(org.openqa.selenium.support.ui.SelectLargeTest)
Mar 26, 2015 3:12:15 AM org.openqa.selenium.testing.JUnit4TestBase$1 starting
INFO: >>> Starting multipleSelectShouldBePossibleIfMulitpleAttributeEmpty(org.openqa.selenium.support.ui.SelectLargeTest)
Mar 26, 2015 3:12:15 AM org.openqa.selenium.testing.JUnit4TestBase$1 finished
INFO: <<< Finished multipleSelectShouldBePossibleIfMulitpleAttributeEmpty(org.openqa.selenium.support.ui.SelectLargeTest)
Mar 26, 2015 3:12:15 AM org.openqa.selenium.testing.JUnit4TestBase$1 starting
INFO: >>> Starting canListDecoratedElements(org.openqa.selenium.support.pagefactory.UsingPageFactoryTest)
Mar 26, 2015 3:12:15 AM org.openqa.selenium.testing.JUnit4TestBase$1 finished
INFO: <<< Finished canListDecoratedElements(org.openqa.selenium.support.pagefactory.UsingPageFactoryTest)
Mar 26, 2015 3:12:15 AM org.openqa.selenium.testing.JUnit4TestBase$1 starting
INFO: >>> Starting testDecoratedElementsShouldBeUnwrapped(org.openqa.selenium.support.pagefactory.UsingPageFactoryTest)
Mar 26, 2015 3:12:15 AM org.openqa.selenium.testing.JUnit4TestBase$1 finished
INFO: <<< Finished testDecoratedElementsShouldBeUnwrapped(org.openqa.selenium.support.pagefactory.UsingPageFactoryTest)
Mar 26, 2015 3:12:15 AM org.openqa.selenium.testing.JUnit4TestBase$1 starting
INFO: >>> Starting canExecuteJsUsingDecoratedElements(org.openqa.selenium.support.pagefactory.UsingPageFactoryTest)
Mar 26, 2015 3:12:15 AM org.openqa.selenium.testing.JUnit4TestBase$1 finished
INFO: <<< Finished canExecuteJsUsingDecoratedElements(org.openqa.selenium.support.pagefactory.UsingPageFactoryTest)
go aborted!
org.apache.tools.ant.BuildException: Test org.openqa.selenium.support.LargeTests failed
```

lukeis commented on java/client/src/org/openqa/selenium/support/pagefactory/ByChained.java:
```diff
@@ -48,10 +48,21 @@ public ByChained(By... bys) {
+        elem = elem.findElement(by);
+      }
+    }
+    assert elem != null; // NoSuchElement should have already been thrown
```
Let's not use "assert" but rather throw an Exception explicitly.

Overall, this logic re-write doesn't maintain the existing functionality. Namely if any of the by's do not find the element, it doesn't throw an Element Not Found exception, but rather continues executing all by's until at least one was found.

I would change your for loop to look more like this:

```
for (By by : bys) {
  List<WebElements> elements = by.findElements(elem == null ? context : elem);
  if (!elements.isEmpty()) {
    elem = elements.get(0);
  }
}
```

alb-i986 replied:
Hi Luke,

Thanks for your feedback.

Uhm actually NoSuchEl may be thrown by

by.findElement(context)
Or
elem.findElement(by)

So it will stop and throw as soon as any one of the bys retrieves no
elements. If I'm not mistaken!

Example.
ByChained(by1, by2, by3)
By1 retrieves 3 elements. I save the first
By2 retrieves nothing. NoSuchEl is thrown by elem.findElement(by).

Let me know how it sounds.

About the assert, no problem, I will change it.

Thanks

Alberto Scotto
http://alb-i986.me
[on Android]
Il 26/mar/2015 15:13 "Luke Inman-Semerau" notifications@github.com ha
scritto:

> In java/client/src/org/openqa/selenium/support/pagefactory/ByChained.java
> https://github.com/SeleniumHQ/selenium/pull/372#discussion_r27221662:
> 
> > -      throw new NoSuchElementException("Cannot locate an element using " + toString());
> > -    return elements.get(0);
> > -    if (bys.length == 0) {
> > -      throw new NoSuchElementException("No Bys were specified in this " + getClass().getSimpleName());
> > -    }
> >   +
> > -    WebElement elem = null;
> > -    for (By by : bys) {
> > -      boolean isFirstIteration = elem == null;
> > -      if (isFirstIteration) {
> > -        elem = by.findElement(context);
> > -      } else {
> > -        elem = elem.findElement(by);
> > -      }
> > -    }
> > -    assert elem != null; // NoSuchElement should have already been thrown
> 
> Let's not use "assert" but rather throw an Exception explicitly.
> 
> Overall, this logic re-write doesn't maintain the existing functionality.
> Namely if any of the by's do not find the element, it doesn't throw an
> Element Not Found exception, but rather continues executing all by's until
> at least one was found.
> 
> I would change your for loop to look more like this:
> 
> for (By by : bys) {
>   List<WebElements> elements = by.findElements(elem == null ? context : elem);
>   if (!elements.isEmpty()) {
>     elem = elements.get(0);
>   }
> }
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/SeleniumHQ/selenium/pull/372/files#r27221662.

lukeis replied:
I was trying to point out the previous code path used by.findElements <- s is important and would continue searching through all by's looking for elements.

Your change makes it so if the first by does not locate an element (or essentially any by in the list) it would throw a not found exception.

The change I proposed keeps the logic in place where it would end up calling all the by's in the chain. 

But the existing code has some odd edge cases by the artifact of the last by in the chain is the only one it really cares about and it can be hard to predict results if intermediate by's don't return any elements.

alb-i986 replied:
Ah I see now! Thanks for the further explanation :)
So I've always been misunderstanding the behaviour of ByChained, damn.
The thing is that by reading the javadoc

> This class will find all DOM elements that matches each of the locators in sequence,

I understood that it behaved like a logical AND of Bys, and as such that it would "fail" if any one of the By failed.
Should it have been "each and every" if it had had the meaning I got? Not sure as I'm not english mothertongue.

I'm wondering if it's only me to have misunderstood or there are (many?) others.
If it's only me, then fine, I can just add a new class with the 'AND' behavior. Also I would like to add a line to the javadoc, explaining the thing, as well as maybe an example, why not.

What do you think?

alb-i986 replied:
Hey
sorry, I'm back.
It looks like the current behavior is actually the way I thought, i.e. if any one of the bys returns no elements, findElements will return an empty list.
Please check out https://github.com/alb-i986/selenium/commit/54d0dc8d72f2643183d7f5bbe96ff727fbae9f93

The key is `elems = newElems;` at the end of each iteration, which overwrites the list that is iterated during each step. The first by that returns no elements will set `elems` to empty list; any subsequent iteration of the outer loop will do nothing, since the inner loop will not be executed. BTW this lead me to another improvement: to [exit the loop as soon as elems is empty](https://github.com/alb-i986/selenium/commit/ae639dab1b3a114ae5ad396bdcc315c4fc1d627c)

I made some more refactorings, including replacing the for-each with a standard for, which is supposed to improve readability.
https://github.com/alb-i986/selenium/blob/d336458e8040e7bf4cbc09afe1232777dbe8c0f1/java/client/src/org/openqa/selenium/support/pagefactory/ByChained.java#L74

Please let me know your thoughts
Thanks

alb-i986 replied:
The question now is whether using `WebElement#findElement` instead of `findElements` is really an improvement.
Because originally I quite blindly assumed that retrieving one element must be more efficient than retrieving more, but actually I don't know..

So let's analyse..
I think we can say that most of the times `WebElement#findElement` will be resolved as `RemoteWebElement#findElement`, right?
So, by following through the calls, the question ends up being whether 
`Command.FIND_CHILD_ELEMENT` and `DriverCommand.FIND_ELEMENT` are more efficient than `DriverCommand.FIND_CHILD_ELEMENTS` and `DriverCommand.FIND_ELEMENTS` (respectively).
I mean, my concern is that even at the lowest level findElement were implemented as `findElements.get(0)`.

lukeis commented:
regarding your local build, not sure what's going on there. But you should be able to use Java 8 or 7 to build / run (but not 6, definitely not 6 ;) )

alb-i986 commented:
About the build, maybe I will try to setup a Linux VM on my Mac.
By the way it would be cool if you could provide a Vagrantfile for hassle free building. If you like the idea I may work on that. I already have something from which to start: see https://github.com/alb-i986/vagrantfiles/tree/master/jenkins

lukeis commented:
You should be able to build Selenium on all platforms (windows, mac, linux). But I did start the effort there using docker once... got it "working" but since we aren't doing native events for firefox anymore, stopped working on it.

p0deje commented:
@alb-i986 I have started working on the same thing though it only supports Firefox at the moment (which was fine for me at that point). Take a look at https://github.com/p0deje/selenium-boxes.

alb-i986 commented:
@p0deje 
I'm now finding travis a very good way to run tests hassle-free.
BTW in my case the build (the tests) was failing because I didn't have Firefox installed (LOL).

alb-i986 commented:
Ok I've just pushed my new changes to this branch.
I made quite a lot of commits, before merging I'll squash.
Thanks

alb-i986 commented:
Darn, I rebased on top of the current master, and looks like the comments made over my original commit got lost.
I expected github to keep the comments despite the different hash..
My apologies

Anyways, I'll try to sum up.

So, the discussion was revolving around the original behavior of findElement
I was saying that to me it looks like the original behavior of `ByChained#findElements` is to return an empty list if *any* one of the bys is empty. I've added a couple of unit tests which prove it.

The key is the assignment `elems = newElems` at the end of each iteration of the outer loop. If any one of the bys finds no elements, then `newElems` will be empty, thus the next iteration `elems` will be empty, and so it will always be until the end of the loop, as the inner loop will not be executed.
This BTW lead me to refactor `findElements` so that it exits the outer loop as soon as the inner loop yields no elements.

Then I was wondering if this refactor to `findElement` is really an improvement.
The question should be whether `DriverCommand.FIND_ELEMENT` is actually more efficient than `DriverCommand.FIND_ELEMENTS`.
My concern being that `DriverCommand.FIND_ELEMENT`'s might be implemented in terms of `DriverCommand.FIND_ELEMENTS`. I'd expect not to, but I don't know.

Hope not to have forgotten anything.
Cheers

alb-i986 commented:
Hey @lukeis 

I've completely re-written this PR.
As you said, it was flawed. At the time I didn't understand what you were trying to say, but after writing and debugging a couple of integration tests, now I do.
I've implemented a recursive algorithm which scans the entire tree until it finds the deepest left-most element.
Also, I took out many of the extra changes I introduced in the PR originally, which added noise and confusion.
Should be fine now.
Please check it out.
Thanks

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/SeleniumHQ/selenium?pullRequest=372) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/SeleniumHQ/selenium?pullRequest=372) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/SeleniumHQ/selenium?pullRequest=372) it.</sub>

diemol commented:
Apologies for the poor handling of this PR.

In general, we are discouraging the use of `PageFactory`, but in any case, lots of people use it. We don't want to make more changes to it, and in any case, changes must ensure backward compatibility is complicated. 

Closing this PR.

alb-i986 commented:
Hey,
> In general, we are discouraging the use of PageFactory
Yeah? Is it reported in the docs?
Thanks

alb-i986 commented:
https://www.selenium.dev/documentation/en/guidelines_and_recommendations/page_object_models/

Here PageFactory is mentioned but not discouraged:

> There are other design patterns that also may be used in testing. Some use a Page Factory for instantiating their page objects [..]

diemol commented:
That's a good point @alb-i986! We'll add a note there soon.

barancev commented:
I think that this PR is safe enough from backward compatibility perspective, it does not break the API and improves internals of PageFactory implementation.

diemol commented:
Ok, that is good enough for me, let's reopen it and get it merged then.

diemol commented:
@alb-i986 Could you please fix the conflicts and then we'll get this in?

alb-i986 commented:
I'm working on this but I'm having troubles with getting the build working.
I've tried requesting for help on IRC but I didn't get any reply.
Hopefully you can help me.
I'm on Ubuntu 18.04.4. I got as far as getting a complaint about the "chalk" node module.
Here are the logs: pastebin.com/hkqGM7gC pastebin.com/FXSfGjkS
Thanks.

alb-i986 commented:
Nevermind, I solved.

alb-i986 commented:
I've just created a new patch #8370 which addresses the doubts which arose here, around the functionality of the ByChained. If you look at the review comments above, you can see the original implementation was pretty obscure.
I thought it made sense to open a different PR and address that first, so as to make things clear.
Please check it out.

diemol commented:
@alb-i986, now that #8370 was merged, would you like to continue with this PR?

diemol commented:
I will close this PR as it went stale, but if there is a wish to continue working on it, we can create a new one or fix the conflicts after reopening it.
